### PR TITLE
Add new mysql column `downtime.duration` & `service_state.host_id`

### PR DESCRIPTION
**Pull request by Yonas-net**

### CODE REVIEW

**Comment by julianbrost** [comment]
Corresponding icinga2 PR: https://github.com/Icinga/icinga2/pull/9061

File: schema/mysql/schema.sql [hunk 1]
```diff
@@ -612,6 +613,7 @@ CREATE TABLE downtime (
   is_in_effect enum('n', 'y') NOT NULL,
   start_time bigint unsigned DEFAULT NULL COMMENT 'Time when the host went into a problem state during the downtimes timeframe',
   end_time bigint unsigned DEFAULT NULL COMMENT 'Problem state assumed: scheduled_end_time if fixed, start_time + flexible_duration otherwise',
+  duration bigint unsigned NOT NULL COMMENT 'Duration of the downtime: When the downtime is flexible, this is the same as flexible_duration otherwise scheduled_end_time - scheduled_start_time',
```
Comment (julianbrost) [inline schema/mysql/schema.sql:616]: According to https://github.com/Icinga/icinga2/issues/9049, this column will be used for sorting. Does just adding that as a column without an index on it already improve things there?

Comment (Yonas-net) [reply schema/mysql/schema.sql:616]: Yeah, it does. But an index for this is added with this one. https://github.com/Icinga/icingadb/pull/397

Comment (julianbrost) [reply schema/mysql/schema.sql:616]: Okay, so then that PR is also blocked by this one as it adds indices on columns only introduced here?

File: pkg/icingadb/v1/downtime.go [hunk 2]
```diff
@@ -19,11 +19,13 @@ type Downtime struct {
...
 	ScheduledStartTime types.UnixMilli `json:"scheduled_start_time"`
 	ScheduledEndTime   types.UnixMilli `json:"scheduled_end_time"`
-	FlexibleDuration   uint64          `json:"flexible_duration"`
+	ScheduledDuration  types.UnixMilli `json:"scheduled_duration"`
 	IsFlexible         types.Bool      `json:"is_flexible"`
+	FlexibleDuration   types.UnixMilli `json:"flexible_duration"`
 	IsInEffect         types.Bool      `json:"is_in_effect"`
 	StartTime          types.UnixMilli `json:"start_time"`
 	EndTime            types.UnixMilli `json:"end_time"`
+	Duration           types.UnixMilli `json:"duration"`
```
Comment (julianbrost) [inline pkg/icingadb/v1/downtime.go:28]: This shouldn't use `UnixMilli` as that's a `time.Time` and thus represents a point in time. Either keep it as `uint64` or introduce something like `DurationMilli` that wraps a `time.Duration` and is represented as milliseconds in JSON/SQL.

File: schema/mysql/schema.sql [hunk 3]
```diff
@@ -278,6 +278,7 @@ CREATE TABLE servicegroup_customvar (
 
 CREATE TABLE service_state (
   id binary(20) NOT NULL COMMENT 'service.id',
+  host_id binary(20) NOT NULL COMMENT 'service.host.id',
```
Comment (julianbrost) [inline schema/mysql/schema.sql:281]: Just use `host.id` in the comment, tables like `comment`/`downtime`/`notification` also do it that way.

**Review by julianbrost — CHANGES_REQUESTED** [review]

**Review by julianbrost — CHANGES_REQUESTED** [review]
https://github.com/Icinga/icinga2/pull/9061 adds a `duration` field to downtime history events but this one does not add the corresponding column to `downtime_history`.

**Review by julianbrost — CHANGES_REQUESTED** [review]
```
2021-11-05T18:15:34.406+0100	FATAL	icingadb	Error 1054: Unknown column 'scheduled_duration' in 'field list'
can't perform "INSERT INTO downtime_history (scheduled_start_time,scheduled_duration,endpoint_id,author,is_flexible,scheduled_end_time,scheduled_by,downtime_id,object_type,host_id,parent_id,entry_time,flexible_duration,cancel_time,comment,environment_id,service_id,trigger_time,cancelled_by,has_been_cancelled,triggered_by_id,end_time,duration,start_time) VALUES (:scheduled_start_time,:scheduled_duration,:endpoint_id,:author,:is_flexible,:scheduled_end_time,:scheduled_by,:downtime_id,:object_type,:host_id,:parent_id,:entry_time,:flexible_duration,:cancel_time,:comment,:environment_id,:service_id,:trigger_time,:cancelled_by,:has_been_cancelled,:triggered_by_id,:end_time,:duration,:start_time) ON DUPLICATE KEY UPDATE cancelled_by = VALUES(cancelled_by),has_been_cancelled = VALUES(has_been_cancelled),cancel_time = VALUES(cancel_time)"
```

**Review by julianbrost — CHANGES_REQUESTED** [review]
Code change in general looks good. Schema migration has the issue that it adds `NOT NULL` columns without settings values for existing rows.

**Comment by lippserd** [comment]
> Code change in general looks good. Schema migration has the issue that it adds `NOT NULL` columns without settings values for existing rows.

I removed the history related changes as the duration columns are not required in the history tables.

Also, I think it's okay not to set values for `NOT NULL` columns as we need a clean database anyway.

**Comment by julianbrost** [comment]
> Also, I think it's okay not to set values for `NOT NULL` columns as we need a clean database anyway.

For the purpose of migrating a dev setup that has data from after #391, it would still be nice to be able to migrate it.

**Comment by lippserd** [comment]
> > Also, I think it's okay not to set values for `NOT NULL` columns as we need a clean database anyway.
> 
> For the purpose of migrating a dev setup that has data from after #391, it would still be nice to be able to migrate it.

True. Done!

**Comment by lippserd** [comment]
@julianbrost Any chance to review?

File: schema/mysql/upgrades/1.0.0-rc2.sql [hunk 4]
```diff
@@ -322,3 +322,16 @@ ALTER TABLE flapping_history
...
+ */
+
+ALTER TABLE downtime
+    ADD COLUMN scheduled_duration bigint unsigned NOT NULL AFTER scheduled_end_time,
+    ADD COLUMN duration bigint unsigned NOT NULL COMMENT 'Duration of the downtime: When the downtime is flexible, this is the same as flexible_duration otherwise scheduled_duration' AFTER end_time,
+    MODIFY COLUMN flexible_duration bigint unsigned NOT NULL AFTER is_flexible;
+UPDATE downtime SET scheduled_duration = scheduled_end_time - scheduled_start_time, duration = (CASE WHEN is_flexible = 'Y' THEN flexible_duration ELSE scheduled_end_time - scheduled_start_time END) WHERE scheduled_duration IS NULL;
+
+ALTER TABLE service_state ADD COLUMN host_id binary(20) NOT NULL COMMENT 'service.host.id' AFTER id;
+UPDATE service_state INNER JOIN service ON service.id = service_state.service_id SET service_state.host_id = service.host_id WHERE service_state.host_id IS NULL;
```
Comment (julianbrost) [inline schema/mysql/upgrades/1.0.0-rc2.sql:337]: Just tested this with MariaDB 10.6.4 and the first statement seems to fill the column with a 0x00...00 value, so the second statement affects no rows (filters for `host_id IS NULL` which is marked `NOT NULL`).

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: schema/mysql/upgrades/1.0.0-rc2.sql [hunk 5]
```diff
@@ -407,3 +407,12 @@ ALTER TABLE eventcommand_customvar
 ALTER TABLE notificationcommand_customvar
     ADD INDEX idx_notificationcommand_customvar_notificationcommand_id (notificationcommand_id, customvar_id),
     ADD INDEX idx_notificationcommand_customvar_customvar_id (customvar_id, notificationcommand_id);
+
+ALTER TABLE downtime
+    ADD COLUMN scheduled_duration bigint unsigned NOT NULL AFTER scheduled_end_time,
+    ADD COLUMN duration bigint unsigned NOT NULL COMMENT 'Duration of the downtime: When the downtime is flexible, this is the same as flexible_duration otherwise scheduled_duration' AFTER end_time,
+    MODIFY COLUMN flexible_duration bigint unsigned NOT NULL AFTER is_flexible;
+UPDATE downtime SET scheduled_duration = scheduled_end_time - scheduled_start_time, duration = (CASE WHEN is_flexible = 'Y' THEN flexible_duration ELSE scheduled_end_time - scheduled_start_time END) WHERE scheduled_duration = 0;
```
Comment (julianbrost) [inline schema/mysql/upgrades/1.0.0-rc2.sql:415]: Must be a lower-case `'y'`, otherwise it will always use the else branch.

**Review by julianbrost — CHANGES_REQUESTED** [review]

**Review by julianbrost — CHANGES_REQUESTED** [review]
Now that I can properly diff again with #407:
```diff
 CREATE TABLE `downtime` (
   `id` binary(20) NOT NULL COMMENT 'sha1(environment.id + name)',
   `environment_id` binary(20) NOT NULL COMMENT 'environment.id',
   `triggered_by_id` binary(20) DEFAULT NULL COMMENT 'The ID of the downtime that triggered this downtime. This is set when creating downtimes on a host or service higher up in the dependency chain using the "child_option" "DowntimeTriggeredChildren" and can also be set manually via the API.',
   `parent_id` binary(20) DEFAULT NULL COMMENT 'For service downtimes, the ID of the host downtime that created this downtime by using the "all_services" flag of the schedule-downtime API.',
   `object_type` enum('host','service') COLLATE utf8mb4_bin NOT NULL,
   `host_id` binary(20) NOT NULL COMMENT 'host.id',
   `service_id` binary(20) DEFAULT NULL COMMENT 'service.id',
   `name_checksum` binary(20) NOT NULL COMMENT 'sha1(name)',
   `properties_checksum` binary(20) NOT NULL COMMENT 'sha1(all properties)',
   `name` varchar(548) COLLATE utf8mb4_bin NOT NULL COMMENT '255+1+255+1+36, i.e. "host.name!service.name!UUID"',
   `author` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL,
   `comment` text COLLATE utf8mb4_bin NOT NULL,
   `entry_time` bigint(20) unsigned NOT NULL,
   `scheduled_start_time` bigint(20) unsigned NOT NULL,
   `scheduled_end_time` bigint(20) unsigned NOT NULL,
   `scheduled_duration` bigint(20) unsigned NOT NULL,
   `is_flexible` enum('n','y') COLLATE utf8mb4_bin NOT NULL,
   `flexible_duration` bigint(20) unsigned NOT NULL,
   `is_in_effect` enum('n','y') COLLATE utf8mb4_bin NOT NULL,
   `start_time` bigint(20) unsigned DEFAULT NULL COMMENT 'Time when the host went into a problem state during the downtimes timeframe',
   `end_time` bigint(20) unsigned DEFAULT NULL COMMENT 'Problem state assumed: scheduled_end_time if fixed, start_time + flexible_duration otherwise',
-  `duration` bigint(20) unsigned NOT NULL COMMENT 'Duration of the downtime: When the downtime is flexible, this is the same as flexible_duration otherwise scheduled_end_time - scheduled_start_time',
+  `duration` bigint(20) unsigned NOT NULL COMMENT 'Duration of the downtime: When the downtime is flexible, this is the same as flexible_duration otherwise scheduled_duration',
   `scheduled_by` varchar(767) COLLATE utf8mb4_bin DEFAULT NULL COMMENT 'Name of the ScheduledDowntime which created this Downtime. 255+1+255+1+255, i.e. "host.name!service.name!scheduled-downtime-name"',
   `zone_id` binary(20) DEFAULT NULL COMMENT 'zone.id',
   PRIMARY KEY (`id`),
   KEY `idx_downtime_is_in_effect` (`is_in_effect`,`start_time`) COMMENT 'Downtime list filtered/ordered by severity',
   KEY `idx_downtime_name` (`name`) COMMENT 'Downtime detail filter'
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=DYNAMIC;
```
```diff
 CREATE TABLE `service_state` (
   `id` binary(20) NOT NULL COMMENT 'service.id',
-  `host_id` binary(20) NOT NULL COMMENT 'host.id',
+  `host_id` binary(20) NOT NULL COMMENT 'service.host.id',
   `service_id` binary(20) NOT NULL COMMENT 'service.id',
   `environment_id` binary(20) NOT NULL COMMENT 'environment.id',
   `properties_checksum` binary(20) NOT NULL COMMENT 'sha1(all properties)',
   `state_type` enum('hard','soft') COLLATE utf8mb4_bin NOT NULL,
   `soft_state` tinyint(3) unsigned NOT NULL,
   `hard_state` tinyint(3) unsigned NOT NULL,
   `previous_hard_state` tinyint(3) unsigned NOT NULL,
   `attempt` tinyint(3) unsigned NOT NULL,
   `severity` smallint(5) unsigned NOT NULL,
   `output` longtext COLLATE utf8mb4_bin DEFAULT NULL,
   `long_output` longtext COLLATE utf8mb4_bin DEFAULT NULL,
   `performance_data` longtext COLLATE utf8mb4_bin DEFAULT NULL,
   `normalized_performance_data` longtext COLLATE utf8mb4_bin DEFAULT NULL,
   `check_commandline` text COLLATE utf8mb4_bin DEFAULT NULL,
   `is_problem` enum('n','y') COLLATE utf8mb4_bin NOT NULL,
   `is_handled` enum('n','y') COLLATE utf8mb4_bin NOT NULL,
   `is_reachable` enum('n','y') COLLATE utf8mb4_bin NOT NULL,
   `is_flapping` enum('n','y') COLLATE utf8mb4_bin NOT NULL,
   `is_overdue` enum('n','y') COLLATE utf8mb4_bin NOT NULL,
   `is_acknowledged` enum('n','y','sticky') COLLATE utf8mb4_bin NOT NULL,
   `acknowledgement_comment_id` binary(20) DEFAULT NULL COMMENT 'comment.id',
   `last_comment_id` binary(20) DEFAULT NULL COMMENT 'comment.id',
   `in_downtime` enum('n','y') COLLATE utf8mb4_bin NOT NULL,
   `execution_time` int(10) unsigned DEFAULT NULL,
   `latency` int(10) unsigned DEFAULT NULL,
   `timeout` int(10) unsigned DEFAULT NULL,
   `check_source` text COLLATE utf8mb4_bin DEFAULT NULL,
   `scheduling_source` text COLLATE utf8mb4_bin DEFAULT NULL,
   `last_update` bigint(20) unsigned DEFAULT NULL,
   `last_state_change` bigint(20) unsigned NOT NULL,
   `next_check` bigint(20) unsigned NOT NULL,
   `next_update` bigint(20) unsigned NOT NULL,
   PRIMARY KEY (`id`),
   UNIQUE KEY `idx_service_state_service_id` (`service_id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=DYNAMIC;
```

But we're getting there! Once these two comments are fixed, that should be it.

**Review by julianbrost — CHANGES_REQUESTED** [review]
I should have been more precise. Asked for it to be changed from `service.host.id` to `host.id` before to stay consistent with other tables that also have `host_id` and `service_id` columns :see_no_evil: https://github.com/Icinga/icingadb/pull/403#discussion_r743560435

**Review by julianbrost — APPROVED** [review]

**Pull request merged**